### DISCUSSION TITLE
Why need to force re render after hydration? when using Material UI with styled-components

### DISCUSSION THREAD

**Discussion by Suedo** [vercel/next.js#12693] [Help]
New to NextJS and static site generation, so apologies if this is a stupid/obvious question.

**TL;DR:**
Without force re-rendering after hydration, styled-component styles vanishes after refresh. Need to rerender using `<StylesProvider key={key} generateClassName={generateClassName}>` instead of `<StylesProvider injectFirst>` . Is this the right way ? Why do I need this ?

**Longer Version:**
I am using Material UI along with Styled Components, and I have a Layout Component like so:

```ts
import React from 'react';
import styled from 'styled-components';
import Header from '../header/header';
import { Box } from '@material-ui/core';

/* eslint-disable-next-line */
export interface LayoutProps {
  children: any;
}

const StyledLayout = styled(Box)`
  max-width: 800px;
  // ...
`;

export const Layout = (props: LayoutProps) => {
  return (
    <>
      <Header />
      <StyledLayout>{props.children}</StyledLayout>
    </>
  );
};

export default Layout;
```

I used the method explained in [Controlling priority](https://material-ui.com/guides/interoperability/#controlling-priority-3) section of Material UI docs, and inserted the  `<StylesProvider injectFirst>` section in my `_app.tsx` file, as shown in the [nestjs + material-ui example](https://github.com/mui-org/material-ui/tree/master/examples/nextjs), but would **still cause the styles introduced in `Layout.tsx` to vanish after a refresh**

This [SO answer](https://stackoverflow.com/a/61141248/2715083) shows the fix being forcing a re-render after hydration through:

```ts
<StylesProvider key={key} generateClassName={generateClassName}>
```

instead of 

```ts 
<StylesProvider injectFirst>
```

And this worked perfectly for me as well. 

So I wanted to understand if this the optimal way of doings things, or maybe I am doing something wrong that makes such a hack a necessity ?

Thanks a lot for you time :)

**Comment by Suedo** [comment]
bump.. someone have any ideas ?

**Reply by MatthiasEg** [reply]
@Suedo I'm facing the same issue. Have you found a solution?

**Reply by Suedo** [reply]
No I did not :( Gave up and took it as it is. Would be nice if someone could give an answer.
